python/ffsim/states/slater.py

```diff
@@ -0,0 +1,244 @@
+            return convert_bitstring_type(
+                strings,
+                BitstringType.INT,
+                bitstring_type,
+                length=len(orbs_a) + len(orbs_b),
+            )
+
+        return convert_bitstring_type(
+            strings_a,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_a),
+        ), convert_bitstring_type(
+            strings_b,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_marginals(
+    rdm: np.ndarray, sample: list[int], empty_orbitals: list[int], marginal: float
+) -> np.ndarray:
+    """Generates the conditional and marginal probabilities for adding a particle.
+
+    This is a step of the autoregressive sampling, and uses Bayes's rule.
+
+    Args:
+        rdm: A Numpy array with the one-body reduced density matrix.
+        pos_array: A numpy array with the positions of the particles.
+        empty_orbitals: A numpy array with the empty orbitals that a new particle
+            may occupy. The sorted union of ``pos_array`` and ``empty_orbitals``
+            must be equal to ``numpy.arange(num_orbitals)``.
+        marginal: The marginal probability associated to having the particles
+            in the positions determined by ``pos_array``.
+    Returns:
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+    marginals = np.zeros(len(empty_orbitals), dtype=float)
+    for i, orbital in enumerate(empty_orbitals):
+        new_sample = sample.copy()
+        new_sample = sample + [orbital]
+        rest_rdm = rdm[np.ix_(new_sample, new_sample)]
+        marginals[i] = np.linalg.det(rest_rdm)
+    return marginals
+
+
+def _autoregressive_slater(
+    rdm: np.ndarray,
+    norb: int,
+    nelec: int,
+    seed: np.random.Generator | int | None = None,
+) -> list[int]:
+    """Autoregressively sample positions of particles for a Slater-determinant wave
+    function using a determinantal point process.
+
+    Args:
+        rdm: A numpy array with the one-body reduced density matrix.
+        norb: Number of orbitals.
+        nelec: Number of electrons.
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+    Returns:
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    probs = np.diag(rdm).real / nelec
+    sample = [rng.choice(norb, p=probs)]
+    marginal = [probs[sample[0]]]
+    all_orbs = set(range(norb))
+    for k in range(nelec - 1):
+        empty_orbitals = list(all_orbs.difference(sample))
```

Comment:
I realize now that this code can be made more efficient. There is actually no need for the `empty_orbitals` variable here. Just update `all_orbs` at the end of the loop by removing the chosen item. But you should actually rename `all_orbs` to `empty_orbitals`, and you need to remove the first item before the start of the loop.